src/include/random.h

```diff
@@ -0,0 +1,391 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+*/
+
+#ifndef CEPH_RANDOM_H
+#define CEPH_RANDOM_H 1
+
+#include <mutex>
+#include <random>
+
+// Basic random number facility, adapted from N3551:
+namespace ceph {
+namespace util {
+
+inline namespace version_2017_0 {
+
+namespace detail {
+
+template <typename T>
+T& urng()
+{
+ static T rng_engine;
+ return rng_engine;
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(const int seed, MutexT& m, EngineT& e)
+{
+ std::lock_guard<MutexT> lg(m);
+ e.seed(seed);
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(MutexT& m, EngineT& e)
+{
+ static std::random_device rd;
+
+ std::lock_guard<MutexT> lg(m);
+ e.seed(rd());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(MutexT& m)
+{
+ randomize_rng(m, detail::urng<EngineT>());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(const int n, MutexT& m)
+{
+ std::lock_guard<MutexT> lg(m);
+ detail::urng<EngineT>().seed(n);
+}
+
```

Comment:
That seems reasonable. In that case my recommendation would be to take the 'nerfed' version of <random> and rework it to use all thread_local variables instead of locks.

Regarding seeding, I would say simply don't have a way to 'seed' the nerfed global random engines, and if someone WANTS to seed something, make them use the random_number_generator class that you have. That way we avoid the risk of some subsystem seeding a global/thread_local RNG with a deterministic value and forgetting to re-randomize it and reset all the distributions it used.